docs/efficient_gas_calculation_algorithm.md

```diff
@@ -0,0 +1,156 @@
+
+2. A basic block _starts_ right before the following instructions:
+   - `JUMPDEST`,
+   - the first instruction in the code,
+   - an instruction directly after an instruction that has ended previous basic block.
+
+A basic block is a shortest sequence of instructions such that a basic block starts before the first instruction and ends after the last.
+## Algorithm
+
+The algorithm for calculating gas and checking stack requirements precomputes
+the values for basic blocks and during execution the checks are done only once per block.
+
+### Collecting requirements for basic blocks
+
+For a basic block we need to collect the following information:
+
+- total base gas required by instructions,
+- the stack height required (the minimum stack height needed to execute 
+  all instructions in the block),
+- ~~the start-to-end stack height change~~,
+- the relative maximum stack height.
+
+This is done as follows:
+
+1. Split code into basic blocks.
+2. For each basic block:
+
+```python
+class Instruction:
+    gas_required = 0
+    stack_required = 0
+    stack_change = 0
+
+class BasicBlock:
+    gas_required = 0
+    stack_required = 0
+    stack_change = 0  # FIXME: We don't have too keep it.
+    stack_max = 0
+
+def collect_basic_block_requirements(basic_block):
+    for instruction in basic_block:
+        basic_block.gas_required += instruction.gas_required
+        
+        current_stack_required = instruction.stack_required - basic_block.stack_change
+        basic_block.stack_required = max(basic_block.stack_required, current_stack_required)
+        
+        basic_block.stack_change += instruction.stack_change
+        
+        basic_block.stack_max = max(basic_block.stack_max, basic_block.stack_change)
+```
+
+### Checking basic block requirements
+
+During execution, before executing an instruction that starts a basic block,
+the basic block requirements must be checked.
+
+```python
+class ExecutionState:
+    gas_left = 0
+    stack = []
+
+def check_basic_block_requirements(state, basic_block):
+    state.gas_left -= basic_block.gas_required
+    if state.gas_left < 0:
+        raise OutOfGas()
+    
+    if len(state.stack) < basic_block.stack_required:
+        raise StackUnderflow()
+    
+    if len(state.stack) + basic_block.stack_max > 1024:
+        raise StackOverflow()
+```
+
+## Misc
+
+### EVM may terminate earlier
+
+Because requirements for a whole basic block are checked up front, the instructions
+that have observable external effects might not be executed although they would be
+executed if the gas counting would have been done per instruction.
+This is not a consensus issue because the execution terminates with a "hard" exception
+anyway (and all effects are reverted) but might produce unexpected traces 
+or terminate with a different exception type.
+
+### Current "gas left" value
+
+In EVMJIT also `GAS` and _call_ instructions begin a basic block. This is because
```

Comment:
just a rewording
```suggestion
In EVMJIT additional instructions that begin a basic block are `GAS` and any of the _call_ instructions. This is because
```